File: exercises/Exercise017.java
public class Exercise017 {

    private static String[] singles = {
            "one", "two", "three", "four", "five","six","seven","eight","nine","ten","eleven","twelve",
            "thirteen","fourteen","fifteen","sixteen","seventeen","eighteen","nineteen"
    };

    private static String[] _tenFolds = {
                    "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
    };

    public static void main(String[] args) {
        int total = 0;
        for (int i = 1; i <= 1000; i++)
            total += letterCount(i);
        System.out.println(total);
    }

    public static int letterCount(int n) {
        if (n < 0 || n > 1000) throw new IllegalArgumentException("Incorrect input, we must have 0 <= n <= 1000");
        if (n == 0) return 0;
        if (n < 20) return singles[n - 1].length();
        if (n < 100)
            return _tenFolds[(n / 10) - 2].length() + letterCount(n % 10);
        if (n < 1000)
        {
            if (n % 100 == 0)
                return singles[n / 100 - 1].length() + "hundred".length();
            return singles[n / 100 - 1].length() + "hundredand".length() + letterCount(n % 100);
        }
        // n == 1000 here
        return "onethousand".length();
    }

}

File: python/exercise017.py
singles = ["one", "two", "three", "four", "five","six","seven","eight","nine","ten","eleven","twelve",
            "thirteen","fourteen","fifteen","sixteen","seventeen","eighteen","nineteen"]
tenfolds = ["twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]


def letter_count(n):
    if n == 0:
        return 0
    if n < 20:
        return len(singles[n-1])
    if n < 100:
        return len(tenfolds[int(n / 10) - 2]) + letter_count(n % 10)
    if n < 1000:
        if n % 100 == 0:
            return len(singles[ int(n / 100) - 1]) + len("hundred")
        return len(singles[int(n / 100) - 1]) + len("hundredand") + letter_count(n % 100)
    return len("onethousand")

total = 0
for i in range(1,1001):
    total += letter_count(i)
print(total)